[MXNET-1255] update hybridize documentation (#13597)
apache/incubator-mxnet: docs/tutorials/gluon/hybrid.md

# Hybrid - Faster training and easy deployment

> *Related Contents:*

Reviewer: nit: Content instead of Contents (but don't restart CI just for that change).

Author: I added a link to the [new dive into deeplearning book](http://en.diveintodeeplearning.org/chapter_computational-performance/hybridize.html) so made it "contents"

Reviewer: It's a semantic thing I guess. There's probably a rule here but... usually you can say Content is plural already (as it refers to a mass of stuff) and you don't usually say "Related Contents". It's more common to say "Related Content" even if there's more than one item in the list.
It is weird because you do say "Table of Contents", not "Table of Content". 
🤷‍♂️ English is weird. It's not a showstopper... it's fine whichever way you want to do it. (but do it my way). Jk

> *Related Contents:*

Reviewer: ```suggestion
*Related Content:*
```

net2 = gluon.SymbolBlock.imports('model-symbol.json', ['data'], 'model-0001.params')
```

> ## Operators that does not work with hybridize

Reviewer: ```suggestion
## Operators that do not work with hybridize
```

> While most APIs are the same in NDArray and Symbol, there are some differences. Writting `F.operator` and call `hybridize` may not work all the time.

Reviewer 2: Nit : ```Writing``` instead of ```writting```

> If you want to hybridize your model, you must use `F.some_operator` in your 'hybrid_forward' function, F will be 'mxnet.nd' before you hybridize and 'mxnet.sym' after hybridize. While most APIs are the same in NDArray and Symbol,  there are some differences. Writing `F.some_operator` and call `hybridize` may not work all the time.

Reviewer: ```suggestion
If you want to hybridize your model, you must use `F.some_operator` in your 'hybrid_forward' function. 
`F` will be 'mxnet.nd' before you hybridize and 'mxnet.sym' after hybridize. While most APIs are the same in NDArray and Symbol,  there are some differences. Writing `F.some_operator` and call `hybridize` may not work all of the time.
```

Here we list some frequently used NDArray APIs that can't be hybridized and provide you the work arounds.

### Element-wise Operators

The following arithmetic and comparison APIs are automatically broadcasted if the input NDArrays have different shapes.

> However, that's not the case in Symbol API, it's not automatically broadcasted and you have to manually specify whether to use element-wise operator or broadcast operators.

Reviewer: ```suggestion
However, that's not the case in Symbol API. It's not automatically broadcasted, and you have to manually specify whether to use element-wise operator or broadcast operators.
```

> | NDArray APIs  | Description  |

Reviewer: This table is a little hard to read. Could the `<=>` be replaced with columns instead?

|---|---|

> | [*NDArray.\__add\__*](https://mxnet.incubator.apache.org/api/python/ndarray/ndarray.html#mxnet.ndarray.NDArray.__add__) | x.\__add\__(y) <=> x+y <=> mx.nd.add(x, y)  |

Reviewer: Just a suggestion: You could try using math formatting for better presentation?

|[*broadcast_not_equal*](https://mxnet.incubator.apache.org/api/python/symbol/symbol.html#mxnet.symbol.broadcast_not_equal) | Returns the result of element-wise *not equal to* (!=) comparison operation with broadcasting. |
|[*broadcast_greater*](https://mxnet.incubator.apache.org/api/python/symbol/symbol.html#mxnet.symbol.broadcast_greater) | Returns the result of element-wise *greater than* (>) comparison operation with broadcasting. |
|[*broadcast_greater_equal*](https://mxnet.incubator.apache.org/api/python/symbol/symbol.html#mxnet.symbol.broadcast_greater_equal) | Returns the result of element-wise *greater than or equal to* (>=) comparison operation with broadcasting. |
|[*broadcast_lesser*](https://mxnet.incubator.apache.org/api/python/symbol/symbol.html#mxnet.symbol.broadcast_lesser) |	Returns the result of element-wise *lesser than* (<) comparison operation with broadcasting. |
|[*broadcast_lesser_equal*](https://mxnet.incubator.apache.org/api/python/symbol/symbol.html#mxnet.symbol.broadcast_lesser_equal) | Returns the result of element-wise *lesser than or equal to* (<=) comparison operation with broadcasting. |

> For example, if you wan to add a NDarray to your input x, use `broadcast_add` instead of `+`:

Reviewer 3: nit: "want"

```python
def hybrid_forward(self, F, x):

>     # avoid writting: return x + F.ones((1, 1))

Reviewer 2: Same nit above :)

    return F.broadcast_add(x, F.ones((1, 1)))
```

### Shape

Gluon's imperative interface is very flexible and allows you to print shape of the NDArray. However, Symbol does not have shape attributes. As a result, you need to avoid printing shapes in `hybrid_forward`.
Otherwise, you will get the following error:
```bash
AttributeError: 'Symbol' object has no attribute 'shape'
```

### Slice

> [] in NDArray is to get a slice from the array. [] in Symbol is to get an output from a grouped symbol.

Reviewer 3: nit : how about 
`[]` in NDArray is used to get a slice from the array. However, `[]` in Symbol is used to get an output from a grouped symbol.

> Gluon's imperative interface is very flexible and allows you to print shape of the NDArray. However, Symbol does not have shape attributes. As a result, you need to avoid printing shapes in `hybrid_forward`.

Reviewer: ```suggestion
Gluon's imperative interface is very flexible and allows you to print the shape of the NDArray. However, Symbol does not have shape attributes. As a result, you need to avoid printing shapes in `hybrid_forward`.
```

For example, you will get different result for the following method before and after hybridization.

```python
def hybrid_forward(self, F, x):
    return x[0]
```

> The current workaround is explicitly call [`slice`](https://mxnet.incubator.apache.org/api/python/ndarray/ndarray.html#mxnet.ndarray.NDArray.slice) operators in `hybrid_forwar`d.

Reviewer 3: nit : `hybrid_forward`

> The current workaround is explicitly call [`slice`](https://mxnet.incubator.apache.org/api/python/ndarray/ndarray.html#mxnet.ndarray.NDArray.slice) operators in `hybrid_forwar`d.

Reviewer 3: nit: _to_ explicitly call

> For example, you will get different result for the following method before and after hybridization.

Reviewer: ```suggestion
For example, you will get different results for the following method before and after hybridization.
```

### Not implemented operators

Some of the often used operators in NDArray  are not implemented in Symbol, and will cause hybridization failure

#### Array API

mx.nd.array() is used a lot but Symbol does not have the array API. The current workaround is to use F.ones/ F.zeros/ F.full which exists in both NDArrays and Symbols.

#### In-Place Arithmetic Operators

> In place arithmetic operators are also used a lot in Gluon imperative mode. You need to avoid that and write the operations explicitly in `hybrid_forward`.

Reviewer 3: nit : In-place

> Some of the often used operators in NDArray  are not implemented in Symbol, and will cause hybridization failure

Reviewer: ```suggestion
Some of the often used operators in NDArray are not implemented in Symbol, and will cause hybridization failure.
```

> For example, avoid `x += y` and use `x  = x + y`, other wise you will get `NotImplementedError`.

Reviewer 3: nit: otherwise

> `mx.nd.array()` is used a lot but Symbol does not have the array API. The current workaround is to use F.ones/ F.zeros/ F.full which exists in both NDArrays and Symbols.

Reviewer: ```suggestion
`mx.nd.array()` is used a lot, but Symbol does not have the `array` API. The current workaround is to use `F.ones`, `F.zeros`, or `F.full`, which exist in both the NDArray and Symbol APIs.
```

> In-place arithmetic operators are also used a lot in Gluon imperative mode. You need to avoid that and write the operations explicitly in `hybrid_forward`.

Reviewer: "avoid that"? What exactly? All in-place operators? Can you clarify this?
Maybe better to say something like...
```suggestion
In-place arithmetic operators may be used in Gluon imperative mode, however if you expect to hybridize, you should write the operations explicitly instead.
```

Author: avoid all in-place operators. I have added example: avoid `x+=y` and use `x = x + y`.

| NDArray in-place arithmetic operators | Description |
|---|---|

> |[*NDArray.\__iadd\__*](https://mxnet.incubator.apache.org/api/python/ndarray/ndarray.html#mxnet.ndarray.NDArray.__iadd__) |	x.__iadd__(y) <=> x+=y |

Reviewer: Notice that the double underscore is bolding the text. I'd be worried what this will be interpreted as by Sphinx/rst down the line.
You can escape it, or like I mentioned before use `:math:`.

    # avoid writing: return x + F.ones((1, 1))
    return F.broadcast_add(x, F.ones((1, 1)))
```

If you used `+`, it would still work before hybridization, but will throw an error of shape missmtach after hybridization.

### Shape

Gluon's imperative interface is very flexible and allows you to print the shape of the NDArray. However, Symbol does not have shape attributes. As a result, you need to avoid printing shapes in `hybrid_forward`.
Otherwise, you will get the following error:
```bash
AttributeError: 'Symbol' object has no attribute 'shape'
```

### Slice
`[]` in NDArray is used to get a slice from the array. However, `[]` in Symbol is used to get an output from a grouped symbol.
For example, you will get different results for the following method before and after hybridization.

```python
def hybrid_forward(self, F, x):
    return x[0]
```

The current workaround is to explicitly call [`slice`](https://mxnet.incubator.apache.org/api/python/ndarray/ndarray.html#mxnet.ndarray.NDArray.slice) or [`slice_axis`](https://mxnet.incubator.apache.org/api/python/ndarray/ndarray.html#mxnet.ndarray.NDArray.slice_axis) operators in `hybrid_forward`.

### Not implemented operators

> Some of the often used operators in NDArray are not implemented in Symbol, and will cause hybridization failure

Reviewer: ```suggestion
Some of the often used operators in NDArray are not implemented in Symbol, and will cause hybridization failure.
```

#### NDArray.asnumpy

> Symbol does not support `asnumpy` function, you need to avoid calling `asnumpy` in `hybrid_forward`.

Reviewer: ```suggestion
Symbol does not support the `asnumpy` function. You need to avoid calling `asnumpy` in `hybrid_forward`.
```
